risc0/circuit/rv32im/src/prove/emu/exec/mod.rs

```diff
@@ -436,8 +436,11 @@ impl<'a, 'b, S: Syscall> Executor<'a, 'b, S> {
     fn ecall_software(&mut self) -> Result<bool> {
         tracing::trace!("[{}] ecall_software", self.insn_cycles);
         let into_guest_ptr = ByteAddr(self.load_register(REG_A0)?);
-        let into_guest_len = self.load_register(REG_A1)? as usize;
-        if into_guest_len > 0 && !is_guest_memory(into_guest_ptr.0) {
+        let into_guest_len = self.load_register(REG_A1)?;
+        if into_guest_len > 0
+            && !(is_guest_memory(into_guest_ptr.0)
+                && is_guest_memory(into_guest_ptr.0 + into_guest_len))
```

Comment:
Note that this needs to be a checked_add

Reply:
I suggested in this PR that we add a `checked_add` to `ByteAddr` and `WordAddr`: https://github.com/risc0/risc0/pull/2713

Reply:
Also, would prefer to use `check_guest_addr` instead of `is_guest_memory` directly, to avoid dropping down the `u32` and staying in `ByteAddr`

Reply:
Note that there is the following check that happens later in this function, which we should remove if we're moving the check up to here:
```
        // The guest uses a null pointer to indicate that a transfer from host
        // to guest is not needed.
        if into_guest_len > 0 && !into_guest_ptr.is_null() {
            Self::check_guest_addr(into_guest_ptr + into_guest_len)?;
            self.store_region(into_guest_ptr, bytemuck::cast_slice(&syscall.to_guest))?
        }
```

Reply:
I've removed these changes.